### PR TITLE
Init JSON multi-sheet extractor plugin

### PR DESCRIPTION
**Pull request by amarbakir**
## Please explain how to summarize this PR for the Changelog:
* Initialize JSON multi-sheet extractor plugin
* Refactor out shared code from plugins/json-extractor into utils/extractor

## Tell code reviewer how and what to test:
The `plugins/json-extractor` and `plugins/json-multisheet-extractor` plugins. Existing behavior has been kept

Requirements:

* Must allow users to upload a single JSON file which results in data correctly uploading to the corresponding sheets
* Should be able to determine sheet to data relationships through top level JSON keys. E.g.
```json
{
	"contacts": [ // sheet 1
		{ // record 1
			"firstName": "John", // header 1, value
            		"lastName": "Doe", // header 2, value
            		"email": "john.doe@example.com"
		},
		{ // record 2
			"firstName": "Jane", // header 1, value
        			"lastName": "Smith",
        			"email": "jane.smith@example.com"
		},
	],
	"orders": [ // sheet 2
		{ // record 1
			"id": "ORD123456", // header 1, value
            		"amount": 250 // header 2, value
		},
	]
}
```

Resulting UI somewhat supports this, but a nice next step would be to allow for a multisheet import in one step through the UI, or at the very least not require one to go back to the files list to reimport the second sheet (load data from sheet `contacts` -> import data to `contacts` & load data from sheet `order` -> import data to `order` all in one step):
Screenshot 2024-10-30 at 6 32 43 PM

### CODE REVIEW

File: plugins/json-extractor/src/parser.ts [hunk 1]
```diff
@@ -13,52 +13,27 @@ export function parseBuffer(buffer: Buffer): WorkbookCapture {
     const parsedData = JSON.parse(fileContents) || []
     const results = Array.isArray(parsedData) ? parsedData : [parsedData]
 
-    // Ensure all items are objects
-    const filteredResults = results.filter(
-      (item) => typeof item === 'object' && item !== null
-    )
+    var sheetCapture = parseSheet(results);
```
Comment (amarbakir) [inline plugins/json-extractor/src/parser.ts]: Addressed https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003531
> Use const instead of var

File: plugins/json-extractor/src/parser.ts [hunk 2]
```diff
@@ -13,52 +13,27 @@ export function parseBuffer(buffer: Buffer): WorkbookCapture {
...
+
+function isEmpty(obj) {
+  for (const prop in obj) {
+    if (Object.prototype.hasOwnProperty.call(obj, prop)) {
+      return false;
+    }
+  }
+
+  return true;
+}
```
Comment (amarbakir) [inline plugins/json-extractor/src/parser.ts:39]: Note here from [Patrick](https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1825993344) and the [AI review bot](https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003540)

Summary of suggestions is to use `return Object.keys(obj).length === 0;`. I considered this as well, but further research showed it to not be performant. See https://stackoverflow.com/questions/679915/how-do-i-test-for-an-empty-javascript-object

File: plugins/json-multisheet-extractor/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
+<!-- START_INFOCARD -->
+
+The `@flatfile/json-multisheet-extractor` plugin parses a JSON file and extracts second-level nested objects into first level defined Sheets in Flatfile.
+
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/README.md:28]: Adding an example of the nesting structure here as per https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1825995964

File: plugins/json-extractor/src/parser.ts [hunk 4]
```diff
@@ -13,52 +13,27 @@ export function parseBuffer(buffer: Buffer): WorkbookCapture {
...
-    })
-
     return {
-      [sheetName]: {
-        headers,
-        data: filteredData,
-      },
+      [sheetName]: sheetCapture,
     } as WorkbookCapture
   } catch (error) {
```
Comment (amarbakir) [inline plugins/json-extractor/src/parser.ts:25]: Bot suggestion: https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003535

Haven't addressed yet. Thoughts?

File: plugins/json-multisheet-extractor/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+<!-- END_INFOCARD -->
+
+> When embedding Flatfile, this plugin should be deployed in a server-side listener. [Learn more](/docs/orchestration/listeners#listener-types)
+
+
+
+## Parameters
+
+#### `options.chunkSize` - `default: "10_000"` - `number` - (optional)
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/README.md]: AI bot suggestion on format https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003560

I took some of it's suggestions, but didn't want to deviate too far from the pattern in this repo

File: plugins/json-multisheet-extractor/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+// Initialize the JSON multi-sheet extractor
+const jsonMultiSheetExtractor = JSONMultiSheetExtractor();
+
+// Register the extractor as a middleware for the Flatfile listener
+listener.use(jsonMultiSheetExtractor);
+
+// When a JSON file is uploaded, the data will be extracted and processed using the extractor's parser.
+```
+
+See a working example in our [flatfile-docs-kitchen-sink](https://github.com/FlatFilers/flatfile-docs-kitchen-sink/blob/main/typescript/extractors/index.ts) Github repo.
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/README.md:103]: Note from Patrick https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1825997824

I note as much in Slack https://aprimetechnology.slack.com/archives/C07T66YBDHB/p1730404293554049

We can make this example repo change when we open a PR into the upstream repo

File: plugins/json-multisheet-extractor/jest.config.cjs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  setupFiles: ['../../test/dotenv-config.js'],
+  setupFilesAfterEnv: [
+    '../../test/betterConsoleLog.js',
+    '../../test/unit.cleanup.js',
+  ],
+  testTimeout: 60_000,
+  globalSetup: '../../test/setup-global.js',
+  forceExit: true,
+  passWithNoTests: true,
+}
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/jest.config.cjs:16]: AI bot suggestion around code coverage requirements https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003527

Sounds great on paper but I don't want to add a new PR constraint when the repo isn't enforcing/using code coverage anywhere else. Great change for a separate PR

File: plugins/json-multisheet-extractor/package.json [hunk 8]
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "@flatfile/plugin-json-multisheet-extractor",
+  "version": "0.1.0",
+  "url": "https://github.com/FlatFilers/flatfile-plugins/tree/main/plugins/json-multisheet-extractor",
+  "description": "A plugin for parsing json files into multiple sheets in Flatfile.",
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/package.json]: `json` to `JSON` https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1825998708

File: plugins/json-multisheet-extractor/package.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  "types": "./dist/index.d.ts",
+  "files": [
+    "dist/**"
+  ],
+  "scripts": {
+    "build": "tsup",
+    "build:watch": "tsup --watch",
+    "build:prod": "NODE_ENV=production tsup",
+    "checks": "tsc --noEmit && attw --pack . && publint .",
+    "lint": "tsc --noEmit",
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/package.json:47]: `eslint` + autofix script idea from Patrick https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826001690

File: plugins/json-multisheet-extractor/src/index.ts
Comment (amarbakir) [inline plugins/json-multisheet-extractor/src/index.ts:1]: Discussion around best practice for/against logic in `index.ts` files https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826006387

We agreed to stick to this repo's pattern for now

File: plugins/json-multisheet-extractor/src/parser.spec.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
+import * as fs from 'fs'
+import * as path from 'path'
+import { parseBuffer } from './parser'
+
+describe('parser', function () {
+  const buffer: Buffer = fs.readFileSync(
+    path.join(__dirname, '../ref/test-basic.json')
+  )
+
+  test('JSON to WorkbookCapture', () => {
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/src/parser.spec.ts:27]: Suggestions on keeping things DRY with respect to the repeated headers in this file https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003564

File: plugins/json-multisheet-extractor/src/parser.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import { WorkbookCapture, parseSheet } from '@flatfile/util-extractor'
+
+
+export function parseBuffer(buffer: Buffer): WorkbookCapture {
+  try {
+    const fileContents = buffer.toString('utf8')
+
+    if (!fileContents) {
+      console.log('Invalid file contents')
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/src/parser.ts]: `console.error` suggestion from Patrick https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826015783

Comment (amarbakir) [reply plugins/json-multisheet-extractor/src/parser.ts]: Also AI suggestions about the same + error handling standardization https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003546

File: plugins/json-multisheet-extractor/src/parser.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+          try {
+              const result = parseSheet(value);
+              results[sheet] = result; // Store result in the results object
+          } catch (error) {
+              console.error(`Error processing sheet "${sheet}":`, error);
+          }
+      } else {
+          console.warn(`Skipping key "${sheet}" - not a string/array pair.`);
+      }
+    });
```
Comment (amarbakir) [inline plugins/json-multisheet-extractor/src/parser.ts:35]: AI initiated thread on error handling https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003554

1. Array type validation is delegated to the parseSheet function
2. Failing fast is not necessarily the best approach. Requires Flatfile product input. Personally, I'd rather pull out as much data as possible and allow for a partial success than a complete failure. We could likely use more nuanced response types to address this (e.g. sheet error is returned as part of the SheetCapture object for surfacing to the user, otherwise they might not know what part of a potentially large JSON file is broken)
3. Will set the type for results to WorkbookCapture as opposed to the type another level below (the AI eventually suggests `Record<string, SheetCapture>`, which is effectively WorkbookCapture)

File: utils/extractor/src/index.ts [hunk 13]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
   data: Flatfile.RecordData[]
   metadata?: { rowHeaders: number[] }
 }
+
+export function parseSheet(jsonArray: any[]): SheetCapture {
```
Comment (amarbakir) [inline utils/extractor/src/index.ts:297]: AI suggested depth limit: https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003570

Deferring for now as this has product implications (see deeply nested test case examples)

File: utils/extractor/src/index.ts [hunk 14]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
...
+      return res
+    }
+
+    // Flatten the first item to determine headers
+    const firstItem = flattenObject(filteredResults[0])
+    const headers = Object.keys(firstItem).filter((header) => header !== '')
+
+    // Flatten and filter all rows
+    const filteredData = filteredResults.map((row) => {
+      const flattedRow = flattenObject(row)
```
Comment (amarbakir) [inline utils/extractor/src/index.ts]: Naming https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826030629

File: utils/extractor/src/index.ts [hunk 15]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
...
+    // Ensure all items are objects
+    const filteredResults = jsonArray.filter(
+      (item) => typeof item === 'object' && item !== null
+    )
+
+    if (filteredResults.length === 0) {
+      return {} as SheetCapture
+    }
+
+    // Custom flatten function
```
Comment (amarbakir) [inline utils/extractor/src/index.ts]: More descriptive comment https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826021390

File: utils/extractor/src/index.ts [hunk 16]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
...
+    const flattenObject = (obj: any, parent: string = '', res: any = {}) => {
```
Comment (amarbakir) [inline utils/extractor/src/index.ts:311]: Suggestion from Patrick to not mutate arguments and instead copy values and return the changed result https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826028605

I didn't want to change this refactored code too much as the goal was simply to move it to make it available to both JSON extractors. However, recursive functions typically tend to leverage this mutation pattern, and it's almost idiomatic. I'll add a comment that notes as much

Note that using a non mutating recursive approach adds significantly complexity to merging the final results, likely making them much slower for large datasets

File: utils/extractor/src/index.ts [hunk 17]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
...
+      for (let key in obj) {
+        const propName = parent ? parent + '.' + key : key
+        if (typeof obj[key] === 'object') {
+          flattenObject(obj[key], propName, res)
+        } else {
+          res[propName] = obj[key]
+        }
+      }
+      return res
+    }
```
Comment (amarbakir) [inline utils/extractor/src/index.ts:321]: AI suggestions around type safety https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003574

Much of this doesn't appear to achieve anything significant. Ignoring the depth check (addressed above) usage of unknown, calling `hasOwnProperty` on keys being pulled from the object itself, and doing a null check before a type check don't add type safety as far as I can tell

The array check is interesting however, as I don't think the initial implementation handled that. Let me see how it behaves for arrays and I'll post back here

Comment (amarbakir) [reply utils/extractor/src/index.ts:321]: Actually part of this Array edge case seems to be handled below when computing the filtered data via `value: Array.isArray(cell) ? JSON.stringify(cell) : cell,`. Will verify. I think it's worth adding some unit tests as well for parseSheet now that it's in the utils package

File: utils/extractor/src/index.ts [hunk 18]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
...
+    const filteredData = filteredResults.map((row) => {
+      const flattedRow = flattenObject(row)
+      return headers.reduce((filteredRow, header) => {
+        const cell = flattedRow[header]
+        filteredRow[header] = {
+          value: Array.isArray(cell) ? JSON.stringify(cell) : cell,
+        }
+        return filteredRow
+      }, {})
+    })
```
Comment (amarbakir) [inline utils/extractor/src/index.ts:337]: AI has some suggestions about value handling https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003577

Will review. Not sure this suggestion adds much more than reading complexity. It seems to change the behavior as well which might not be desirable

File: utils/extractor/src/index.ts [hunk 19]
```diff
@@ -293,3 +293,53 @@ export type SheetCapture = {
...
+
+    return {
+      headers,
+      data: filteredData,
+    } as SheetCapture;
+  } catch (error) {
+    console.error('An error occurred:', error)
```
Comment (amarbakir) [inline utils/extractor/src/index.ts:344]: AI suggests improving error handling for sensitive data: https://github.com/FlatFilers/flatfile-plugins/pull/693/files#r1826003566

Will surface to the maintaining team, as this could hide critical debugging info. End users might already have agreed to have this kind of data logged

**Review by PFacheris — APPROVED** [review]
Looks great! Thanks for the detailed responses.

File: plugins/json-extractor/src/parser.ts (hunk 4 above)
Comment (amarbakir) [reply plugins/json-extractor/src/parser.ts:25]: Ignoring for now

File: utils/extractor/src/index.ts (hunk 17 above)
Comment (amarbakir) [reply utils/extractor/src/index.ts:321]: Added unit tests that indicate these checks aren't necessary and/or already exist more succinctly elsewhere. Product approach here appears to be that array items are split into columns. E.g.
```json
"Nicknames": [
      "Dr. T",
      "Señior Lamb"
],
```
turns into these sheet rows:
```json
"Nicknames.0": { "value": "Dr. T" },
"Nicknames.1": { "value": "Señior Lamb" },
```

If an array is empty, the resulting column is null. E.g.
```
"Nicknames.0": { "value": undefined },
```

The test added to the `utils/extractor` package captures this

File: utils/extractor/src/index.ts (hunk 18 above)
Comment (amarbakir) [reply utils/extractor/src/index.ts:337]: After reviewing, I think we can defer this for now as well
